### PR TITLE
doc/doxygen: add build-system-basics.md

### PR DESCRIPTION
**Pull request by fjmolinas**
### Contribution description

I've tried to resume some of the advice I've seen @cladmi @jcarrano and other give on `Makefile` variable declaration/evaluation. 

### Testing procedure

- Have I made a technical mistake int he description?

- Does naming make sense?

- Does the placing make sense?

### Issues/PRs references

Part of #8420

### CODE REVIEW

**Comment by fjmolinas** [comment]
@cladmi I know you are quite busy but If you manage to scheme threw this at some point It would be highly appreciated :)

**Comment by jcarrano** [comment]
That's pretty good. Maybe you could add a link to make's documentation too- my impression is that most people never have never read the official manual.

File: doc/doxygen/src/build-system-basics.md [hunk 1]
```diff
@@ -0,0 +1,56 @@
+Build System Basics                                       {#build-system-basics}
+============================
+
+Variables declaration guidelines              {#variable-declaration-guidelines}
+============
+
+export (export OUTPUT = $(shell some-command))
```
Comment (smlng) [inline doc/doxygen/src/build-system-basics.md]: compiled to HTML these headlines look a bit ugly, I would move the sample command below into the section body.

Comment (fjmolinas) [reply doc/doxygen/src/build-system-basics.md]: addressed

File: doc/doxygen/src/build-system-basics.md [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+--------------------
+
+  Exporting a variable means it will be evaluated on every `shell` call which slows
+  down the build system so avoid exporting a variable if unneeded. 
+
+  If an export is actually needed by a `sub-make` then export the variable only for
+  the needed targets using `target-export-variables (more in `makefiles/utils/variables.mk`)
```
[the remark this answers is not in the document]
Comment (fjmolinas) [reply doc/doxygen/src/build-system-basics.md]: addressed

**Comment by fjmolinas** [comment]
> To me it would make sense to have one build-system page which combines these newly added basics and the existing tricks page. 

Should I do this in this PR as well?

**Comment by smlng** [comment]
not sure, @miri64 what do you think: add first, cleanup later - or directly in this PR?! It definitely makes no sense (to me) to have 2 build system pages at the top level of the docu navigation (menu), so this must be addressed.

**Comment by smlng** [comment]
sry. This PR adds (as per title) some basic infos on the build system, however looking at the current docu we already have http://api.riot-os.org/advanced-build-system-tricks.html these two would sit side-by-side. However this doesn't make sense and I suggest(ed) to introduce a general *Build System* info page and make the 2 subpages of that.

Hope this was understandable - kind of...

**Comment by miri64** [comment]
In my opinion, I don't see any reason to not have two pages. One with more basic stuff and one with more advanced insights. These, to my mind, are addressed to two different audiences, so I think having two documents makes perfect sense. We kind of already do this. We have the "quickest start" page and then a lot of more specific pages of in-depth details.

File: doc/doxygen/src/build-system-basics.md [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+the needed targets using `target-export-variables` (more in
+`makefiles/utils/variables.mk`)
+
+[gnumake doc](https://www.gnu.org/software/make/manual/html_node/Variables_002fRecursion.html)
+
+Use memoized for variables referencing a function/command
+--------------------------
+
+```
+recursively expanded variable: OUTPUT = $(shell some-command)
```
[the remark this answers is not in the document]
Comment (fjmolinas) [reply doc/doxygen/src/build-system-basics.md]: `recursive` is the term used in the doc

> The first flavor of variable is a recursively expanded variable.

https://www.gnu.org/software/make/manual/html_node/Flavors.html

File: doc/doxygen/src/build-system-basics.md [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+Exporting a variable means it will be evaluated on every `target` call which slows
+down the build system. Always avoid exporting a variable if unneeded. 
+
+If an export is actually needed by a `sub-make` then export the variable only for
+the needed targets using `target-export-variables` (more in
+`makefiles/utils/variables.mk`)
+
+Exported variables ("global variable") are hard to remove, specially when badly
+documented. If no one knows why its there and no one knows where it can be used
+then no one knows if its safe to remove since its present for every target.
```
Comment (waehlisch) [inline doc/doxygen/src/build-system-basics.md]: s/its/it is/
That the reason why global variables need clear documentation.

File: doc/doxygen/src/build-system-basics.md [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+[gnumake doc](https://www.gnu.org/software/make/manual/html_node/Variables_002fRecursion.html)
+
+## Use memoized for variables referencing a function/command
```
Comment (waehlisch) [inline doc/doxygen/src/build-system-basics.md]: s/function\/command/function or command/

File: doc/doxygen/src/build-system-basics.md [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+
+### recursively expanded variable:
+
+~~~~~~~~~~~~~~~~
+OUTPUT = $(shell some-command $(ANOTHER_VARIABLE))
+~~~~~~~~~~~~~~~~
+
+- When using `=` the value of the variable is only declared, but not set,
+  therefore will only be evaluated when expanded (used) somewhere in the
```
Comment (waehlisch) [inline doc/doxygen/src/build-system-basics.md]: therefore the variable will ...

File: doc/doxygen/src/build-system-basics.md [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+  makefile. Then if `$(OUTPUT)` is never expanded, `some-command`
```
Comment (waehlisch) [inline doc/doxygen/src/build-system-basics.md]: We don't need the "then".

File: doc/doxygen/src/build-system-basics.md [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+  is never executed and `ANOTHER_VARIABLE` not expanded.
+
+- They will be evaluated every time the variable is expanded. In the example
```
Comment (waehlisch) [inline doc/doxygen/src/build-system-basics.md]: who is "they"?

File: doc/doxygen/src/build-system-basics.md [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+~~~~~~~~~~~~~~~~
+OUTPUT := $(shell some-command $(ANOTHER_VARIABLE))
+~~~~~~~~~~~~~~~~
+
+- When using `:=` the value is only expanded once, expanding any reference to
+  other variables or functions. If `OUTPUT` is always used at least once and
+  evaluates a costly function (`some command`) then use `:=`.
+
+- When using `:=` the variable will be evaluated even if not needed, which is not
+  great if `some command` or functions evaluated by `ANOTHER_VARIABLE` are slow.
```
Comment (waehlisch) [inline doc/doxygen/src/build-system-basics.md]: s/is not great/introduces unnecessary delay, in particular/

File: doc/doxygen/src/build-system-basics.md [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+  It can also cause a failure in a worse case scneario (think what happens if a
```
Comment (waehlisch) [inline doc/doxygen/src/build-system-basics.md]: s/worse case/worst-case/

File: doc/doxygen/src/build-system-basics.md [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+  tool is defined with `:=` but you don't have the tool and you don't need it either)
```
Comment (waehlisch) [inline doc/doxygen/src/build-system-basics.md]: either).

File: doc/doxygen/src/build-system-basics.md [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+
+- The value of variables defined with `:=` depend on the order they are defined.
```
Comment (waehlisch) [inline doc/doxygen/src/build-system-basics.md]: s/value/values/
s/defined with/defined by/

maybe: depend on the order of definition.

File: doc/doxygen/src/build-system-basics.md [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+
+### memoized
```
Comment (waehlisch) [inline doc/doxygen/src/build-system-basics.md]: memoized:

File: doc/doxygen/src/build-system-basics.md [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+
+~~~~~~~~~~~~~~~~
+OUTPUT = $(call memoized,OUTPUT,$(shell some-command))
+~~~~~~~~~~~~~~~~
+
+- Use to get best of both worlds (`=` and `:=`). First evaluation will be as a
```
Comment (waehlisch) [inline doc/doxygen/src/build-system-basics.md]: Don't understand the first sentence. Do you mean: To get best of both worlds, use `=` and `:=`.

Comment (waehlisch) [inline doc/doxygen/src/build-system-basics.md]: The first ... will be considered as

File: doc/doxygen/src/build-system-basics.md [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+  deferred (or recursive) variable but further evaluations will use an immediate
+  value, so `some-command` would be execute once or not at all.  (more in
```
Comment (waehlisch) [inline doc/doxygen/src/build-system-basics.md]: once or not at all (more ...).

File: doc/doxygen/src/build-system-basics.md [hunk 16]
```diff
@@ -0,0 +1,85 @@
...
+ `makefiles/utils/variables.mk`)
+
+- `memoized` is a RIOT function, so don't expect to find it in make doc.
+
+[gnumake doc](https://www.gnu.org/software/make/manual/html_node/Flavors.html)
+
+## more doc
```
Comment (waehlisch) [inline doc/doxygen/src/build-system-basics.md]: s/more doc/Additional documentation/

File: doc/doxygen/src/build-system-basics.md [hunk 17]
```diff
@@ -0,0 +1,85 @@
...
+
+- Deferred vs simple expansion: http://make.mad-scientist.net/deferred-simple-variable-expansion/
```
Comment (waehlisch) [inline doc/doxygen/src/build-system-basics.md]: vs.

**Comment by waehlisch** [comment]
@smlng, I agree with @miri64 . any other reason not to progress?

File: doc/doxygen/src/build-system-basics.md (hunk 17 above)
Comment (fjmolinas) [reply doc/doxygen/src/build-system-basics.md]: done.

File: doc/doxygen/src/build-system-basics.md (hunk 16 above)
Comment (fjmolinas) [reply doc/doxygen/src/build-system-basics.md]: done.

File: doc/doxygen/src/build-system-basics.md (hunk 14 above)
Comment (fjmolinas) [reply doc/doxygen/src/build-system-basics.md]: No, I mean to get the best of both worlds use `memoized`. I'll rephrase let me know if its clearer.

File: doc/doxygen/src/build-system-basics.md (hunk 8 above)
Comment (fjmolinas) [reply doc/doxygen/src/build-system-basics.md]: Thanks, I think this remaned from an earlier re-phrasing, made they explicit.

**Comment by fjmolinas** [comment]
> @fjmolinas, thanks! Very helpful.

Thanks for the comments, addressed!

File: doc/doxygen/src/build-system-basics.md [hunk 18]
```diff
@@ -0,0 +1,87 @@
...
+- The values of variables declared with `:=` depend on the order of definition.
+
+### memoized:
+
+~~~~~~~~~~~~~~~~
+OUTPUT = $(call memoized,OUTPUT,$(shell some-command))
+~~~~~~~~~~~~~~~~
+
+- `memoized` is a RIOT defined function that combines the both worlds (`=` and `:=`).
+  The first time the variable is expanded it will be as a deferred (or recursive)
```
Comment (waehlisch) [inline doc/doxygen/src/build-system-basics.md]: will be used as a ...

File: doc/doxygen/src/build-system-basics.md (hunk 14 above)
Comment (waehlisch) [reply doc/doxygen/src/build-system-basics.md]: function that combines both `=` and `:=`.

File: doc/doxygen/src/build-system-basics.md (hunk 18 above)
Comment (fjmolinas) [reply doc/doxygen/src/build-system-basics.md]: Addressed

File: doc/doxygen/src/build-system-basics.md (hunk 14 above)
Comment (fjmolinas) [reply doc/doxygen/src/build-system-basics.md]: addressed

**Comment by tcschmidt** [comment]
@smlng does this change request still need to be kept? It looks like there is some consensus to proceed with these documents.

**Comment by aabadie** [comment]
Please fix the trailing white space and we are good ;)

**Comment by aabadie** [comment]
(and squash directly)

**Comment by aabadie** [comment]
All green, go!

**Pull request merged**